File: algorithm_data_structure/leetcode/543.diameter-of-binary-tree.py
#
# @lc app=leetcode id=543 lang=python3
#
# [543] Diameter of Binary Tree
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.max_node_num = 1

        def depth(node):
            if not node:
                return 0
            ld = depth(node.left)
            rd = depth(node.right)
            self.max_node_num = max(self.max_node_num, ld + rd + 1)
            return max(ld, rd) + 1

        depth(root)
        return self.max_node_num - 1


# @lc code=end

File: algorithm_data_structure/leetcode/543.diameter-of-binary-tree.cpp
/*
 * @lc app=leetcode id=543 lang=cpp
 *
 * [543] Diameter of Binary Tree
 */

// @lc code=start

#include <algorithm>
#include <vector>

using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    int diameter = 0;

    int getMaxDepth(const TreeNode *root)
    {
        if (root == nullptr)
        {
            return 0;
        }
        int left_depth = getMaxDepth(root->left);
        int right_depth = getMaxDepth(root->right);
        // Update the current diameter at each recursion.
        diameter = max(diameter, left_depth + right_depth);

        return 1 + max(left_depth, right_depth);
    }

    int diameterOfBinaryTree(TreeNode *root)
    {
        diameter = 0;
        getMaxDepth(root);
        return diameter;
    }
};
// @lc code=end
